src/main/java/de/tum/in/www1/artemis/service/connectors/ldap/LdapAuthenticationProvider.java

```diff
@@ -0,0 +1,180 @@
+        User user = getOrCreateUser(authentication, false);
+        if (user != null) {
+            return new UsernamePasswordAuthenticationToken(user.getLogin(), user.getPassword(), user.getGrantedAuthorities());
+        }
+        return null;
+    }
+
+    @Override
+    public User getOrCreateUser(Authentication authentication, String firstName, String lastName, String email, boolean skipPasswordCheck) {
+        // NOTE: firstName, lastName, email is not needed in this case since we always get these values from Jira
+        return getOrCreateUser(authentication, skipPasswordCheck);
+    }
+
+    private User getOrCreateUser(Authentication authentication, Boolean skipPasswordCheck) {
+        String username = authentication.getName().toLowerCase();
+        String password = authentication.getCredentials().toString();
+
+        long start = System.nanoTime();
+
+        final var optionalUser = userRepository.findOneWithGroupsAndAuthoritiesByLogin(username);
+        if (optionalUser.isPresent() && optionalUser.get().isInternal()) {
+            // User found but is internal. Skip external authentication.
+            return null;
+        }
+
+        log.info("Finished userRepository.findOneWithGroupsAndAuthoritiesByLogin in {}", TimeLogUtil.formatDurationFrom(start));
+        start = System.nanoTime();
+
+        // If the following code is executed, the user is either not yet existent or an external user
+
+        final LdapUserDto ldapUserDto = ldapUserService.findByUsername(username).orElseThrow(() -> new BadCredentialsException("Wrong credentials"));
+
+        log.info("Finished ldapUserService.findByUsername in {}", TimeLogUtil.formatDurationFrom(start));
+        start = System.nanoTime();
+
+        // If we want to skip the password check, we can just use the ADMIN auth, which is already injected in the default restTemplate
+        // Otherwise, we create our own authorization and use the credentials of the user.
+        if (!skipPasswordCheck) {
+            byte[] passwordBytes = Utf8.encode(password);
+            boolean passwordCorrect = ldapTemplate.compare(ldapUserDto.getUid().toString(), "userPassword", passwordBytes);
+            log.debug("Compare password with LDAP entry for user " + username + " to validate login");
+            // this is the normal case, where the password is validated
+            if (!passwordCorrect) {
+                throw new BadCredentialsException("Wrong credentials");
+            }
+        }
+
+        log.info("Finished ldapTemplate.compare password in {}", TimeLogUtil.formatDurationFrom(start));
+
+        return optionalUser.orElseGet(() -> {
+            User newUser = userCreationService.createUser(ldapUserDto.getUsername(), null, null, ldapUserDto.getFirstName(), ldapUserDto.getLastName(), ldapUserDto.getEmail(),
+                    ldapUserDto.getRegistrationNumber(), null, "en", false);
+            // TODO: is it ok if a new user does not have any groups?
```

Comment:
From my understanding this should be ok, since the new user is not part of any course yet. Once joining a course, they should automatically get the student-role.

I can only think of this creating issues in case of a tutorial course which everyone joins per default. Then the token after logging in has no groups, but they should be able to directly use the tutorial course. The student might neeeds to login again to use the course in this case.

Reply:
I agree, I don't think this should be a big problem. When an internal user is created, the user seems to also not get any groups